### PR TITLE
Added `MANAGE_EXTERNAL_STORAGE` permission for Android 11 and later

### LINKED ISSUE #2694: Add `MANAGE_EXTERNAL_STORAGE` permission
**Is your feature request related to a problem? Please describe.**
In android 11 we have to add `MANAGE_EXTERNAL_STORAGE`  in order to access other zim files that are present other than our app(outside the app).

**Describe the solution you'd like**
Add this permission

**Additional context**
Previously we discussed: [here](https://github.com/kiwix/kiwix-android/pull/2660#issuecomment-846469934) [here](https://github.com/kiwix/kiwix-android/pull/2660#issuecomment-846536630) [here](https://github.com/kiwix/kiwix-android/pull/2660#issuecomment-846610476) [here](https://github.com/kiwix/kiwix-android/issues/2646)
However, in the Play Store, we will need to justify the reason for requesting that permission.
CC @kelson42

### LINKED ISSUE #2716: Kiwix can't find my ZIM files
Hi, before updating to version 3.4.5 beta, I had a couple of ZIM files in a folder called <Kiwix>, located in the root folder of the internal storage of my phone. After the update, Kiwix can't find them and my library list remains empty. I cleared the cache, with no result.

- Version of Kiwix Android : 3.4.5 beta. 
- Device : Xiaomi Mi 9T pro. 
- OS version : Android 11.

### PR DESCRIPTION
**Pull request by gouri-panda**
Fixes #2694 
Fixes #2716

**Description**
Added `MANAGE_EXTERNAL_STORAGE` permission. For the first time android 11 and later we'll ask specifically to give us permission in order to access all the zim files in android. If the user clicks yes then we'll navigate to Our app's android settings screen, then the user will give access permission.

If the user currently doesn't want to give the permission, We implemented a preference in our app's preference screen. That will allow users to go to our app's navigate settings screen. 

NOTE:-
Currently, if the user first allows the permission and then disallows the permission then the app will show all external files too even though clicking the zim files will not open that. In order to open(read or delete the external)zim files that you have to give the app permission.
It shows on the custom app Which we don't need that permission. (We should really think how to move that)

**ScreenShot**
Screen Shot 2021-10-28 at 18 26 12
Screen Shot 2021-10-28 at 18 15 39
Screen Shot 2021-10-28 at 18 16 16

### CODE REVIEW

**Comment by kevinmcmurtrie** [comment]
Success on Xperia 1 III running Android 11. Torrented to microSd card then this debug build of Kiwix found the files.

**Comment by kelson42** [comment]
@gouri-panda I guess then this PR is ready to merged once this will be code reviewed. I have rebased on `develop` to get a better git history once this is merged. Please in the future secure your PR which are ready to merge are rebased on `develop`.
@abdulwd @mhutti1 @s-ayush2903 Would you please have a look to this PR. Once this is merged, we should probably consider to make a 3.4.6 with this.

**Comment by abdulwd** [comment]
@gouri-panda can you please resolve the todo comment?

**Comment by gouri-panda** [comment]
@abdulwd The `CorePrefsFragment` in java. I thought When we refactor to kotlin, then we can move this extension function to one place. What do you think?

**Comment by kelson42** [comment]
@abdulwd Is your latest comment a blocker to merge this PR to your opinion? Do you see anything else?

**Comment by bryc** [comment]
> @s-ayush2903 @vivadavid @anaxonda @hunterirving @rfc2822 @itlabitto @tibo1902 @l-one1 @mhutti1 @kevinmcmurtrie @virtadpt @Moranic @shapeshifter910 @R2RT Would you please be able to test if the problem of external SD card access on Android11 is fixed with this PR? Here is a test APK https://www.dropbox.com/s/rc2a8z2qt44wc2n/dev-universal-debug.apk?dl=1

This version still does not detect a downloaded zim file, `wikipedia_en_all_maxi_2020-08.zim`, which is 91.9 GB in size, on an exFAT  SD card. I'm using a brand new factory Samsung Galaxy Tab A7 running Android 11. I've followed all the steps here, allowed access in the prompt and in its menu, and it's still stuck in the library, 'no files here'. I've tried renaming the file, reinstalling it multiple times, nothing seems to work. Kiwix on Windows desktop loads this file just fine, so I don't understand what the problem is.  Any thoughts?

**Comment by itlabitto** [comment]
> > @s-ayush2903 @vivadavid @anaxonda @hunterirving @rfc2822 @itlabitto @tibo1902 @l-one1 @mhutti1 @kevinmcmurtrie @virtadpt @Moranic @shapeshifter910 @R2RT Would you please be able to test if the problem of external SD card access on Android11 is fixed with this PR? Here is a test APK https://www.dropbox.com/s/rc2a8z2qt44wc2n/dev-universal-debug.apk?dl=1
> 
> This version still does not detect a downloaded zim file, `wikipedia_en_all_maxi_2020-08.zim`, which is 91.9 GB in size, on an exFAT SD card. I'm using a brand new factory Samsung Galaxy Tab A7 running Android 11. I've followed all the steps here, allowed access in the prompt and in its menu, and it's still stuck in the library, 'no files here'. I've tried renaming the file, reinstalling it multiple times, nothing seems to work. Kiwix on Windows desktop loads this file just fine, so I don't understand what the problem is. Any thoughts?

A couple of questions to understand better,
Did you uninstall the official kiwix app and installed the new kiwix test version?
If you try to open the zim file in the SD card, is it showing the open with Kiwix prompt?
Are you able to see and open other files like simple .jpg or .txt files in that card?

**Comment by bryc** [comment]
> A couple of questions to understand better, 

> Did you uninstall the official kiwix app and installed the new kiwix test version?

Yes, I uninstalled the official app before using this test version. 

> If you try to open the zim file in the SD card, is it showing the open with Kiwix prompt?

Trying to open the zim file using Samsung My Files app doesn't seem to work, seems stuck in a 'processing request' attempt, although Yes, opening it with other apps, such as Amaze and File Manager Plus does give a prompt to open with Kiwix, but after opening it, Kiwix is stuck in the library section, displaying 'no files here' and loading icon.

> Are you able to see and open other files like simple .jpg or .txt files in that card?

Yes, other files open perfectly fine from that SD card on Samsung My Files, and other apps routinely access other files such as video, music, pdf files, etc., from that SD card with no problem.

**Comment by virtadpt** [comment]
> This version still does not detect a downloaded zim file, `wikipedia_en_all_maxi_2020-08.zim`, which is 91.9 GB in size, on an exFAT SD card. I'm using a brand new factory Samsung Galaxy Tab A7 running Android 11. I've followed all the steps here, allowed access in the prompt and in its menu, and it's still stuck in the library, 'no files here'. I've tried renaming the file, reinstalling it multiple times, nothing seems to work. Kiwix on Windows desktop loads this file just fine, so I don't understand what the problem is. Any thoughts?

I'm seeing the same thing over here.

**Review by abdulwd — APPROVED** [review]

**Comment by benkaiser** [comment]
I tried the apk out on my Samsung tablet the other day and experienced the same issue as @bryc . Basically opening from the external storage (sd card) wasn't working. I tried a couple of different file managers, but all failed to open in Kiwix.

Feel free to merge this PR if it resolves for some other devices, but there are definitely some cases where this large file on the external sd card isn't working (maybe it tries to read the entire file off the sd card before opening in the app?).

**Comment by gouri-panda** [comment]
@kelson42 Do you think it's the old samsung issue?

**Comment by kevinmcmurtrie** [comment]
When it's failing, what directory are the files in?  App private storage in the `Android` directory is so slow starting in Android 11 that many apps hang or appear to not work at all.  The workaround is to put the app's files in a different top-level directory then give the app permission to read that.

I put 115 GB of big files in `/storage/hex-codes/Kiwix` and it works very well.
`/storage/hex-codes/Android/data/org.kiwix.kiwixmobile/files` might not work because Google wants you to buy cloud storage rather than insert a cheap microSD card.  If I put OsmAnd+ maps in the standard app directory, it runs about 10000 times slower.

**Comment by benkaiser** [comment]
# #
Screenshot_20211104-092602_My Files
Screenshot_20211104-092626_ReadEra

Here are the screenshots I took trying to open the file. You can see it is on the external SD card under the downloads folder. The downloading UI hangs indefinitely. This is using the official files app, different files apps fail in different ways. Let me know if you would like me to try a specific one.

**Comment by virtadpt** [comment]
> When it's failing, what directory are the files in? App private storage in the `Android` directory is so slow starting in Android 11 that many apps hang or appear to not work at all. The workaround is to put the app's files in a different top-level directory then give the app permission to read that.

SD Card -> Android -> data -> org.kiwix.kiwixmobile -> files -> Kiwix

The file in question is wikipedia_en_all_maxi.zim, dated 4 June 2021, 88.08 gigabytes in size.

Interestingly, if I open a local copy of the Arch Linux wiki, same location on my phone, archlinux_en_all_maxi_2021-05.zim, dated 3 June 2021, 1.04 megabytes in size, it works.  Could it be a size issue?

**Comment by hunterirving** [comment]
I've been testing with the debug apk as well, but haven't been able to open files from an sd card yet. It might be worth mentioning that the device I'm using is running Android 8.1.
Like @virtadpt , I'm also trying to get Kiwix to recognize a large file - "wikipedia_en_all_maxi_2021-03.zim" (82 gigabytes).

**Comment by bryc** [comment]
@benkaiser @virtadpt @hunterirving

Are you able to download the full Wikipedia with pictures, or any other content larger than 4 GB through the Download section of the app? For me, the option is completely greyed out, and it is stuck 'checking' if 4 GB file support is present. I suspect that this may be related to the reason why it can't open large files from SD cards.

**Comment by gouri-panda** [comment]
@bryc The download section isn't related to this issue. The reason it may be greyed out because your SDcard maybe  formatted with FAT32.

**Comment by bryc** [comment]
> @bryc The download section isn't related to this issue. The reason it may be greyed out because your SDcard maybe formatted with FAT32.

I said in my initial post that these issues occur with exFAT.

Perhaps 2354 should be reopened?

**Comment by kelson42** [comment]
There is obviously other bugs/problems related to files on external SD card. This explain to me the other bug reports/remarks reported here. #2727 tend to gather such cases, so I think we sould keep that one open.

**Pull request merged**
